File: hjkim/Week1/01.03/PGS_피로도.java
class Solution {
    static boolean[] visited;
    static int count;

    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(0, k, dungeons);
        return count;
    }

    public void dfs(int depth, int k, int[][] dungeons) {
        for (int i = 0; i < dungeons.length; i++) {
            // 방문하지 않고, 최소 필요 피로도 보다 클 때
            if (!visited[i] && k >= dungeons[i][0]) {
                visited[i] = true;
                dfs(depth + 1, k - dungeons[i][1], dungeons);
                visited[i] = false;
            }
        }
        count = Math.max(count, depth);
    }
}

File: raeunlee/Week1/01.03/PGS_피로도.py
# 과거 코드 참조 - 순열, max
from itertools import permutations

def solution(k, dungeons):
    answer = -1
    
    for p in permutations(dungeons, len(dungeons)):
        tmp = k # 현재 체력
        count = 0
        
        for i in p:
            if tmp >= i[0]:
                tmp -= i[1]
                count += 1
                
        answer = max(answer, count) #기존의 값과 현재 카운트 중에 더 큰것을 지정
                

    return answer